### PR TITLE
bitswap/workers: fix proc / ctx wiring

### PR DESCRIPTION
**Pull request by jbenet**
This commit changes the order of the proc/ctx wiring, to ensure
that the proc has been setup correctly before exiting.

License: MIT
Signed-off-by: Juan Batiz-Benet <juan@benet.ai>

### CODE REVIEW

**Comment by jbenet** [comment]
@whyrusleeping i fixed that proc bug

**Comment by whyrusleeping** [comment]
tests???

**Comment by jbenet** [comment]
@whyrusleeping in the prev config, it's an odd race. i dont know of a reliable way to repro other than run travis a ton of times.

**Comment by jbenet** [comment]
(happy to add something if you have a good suggestion)

**Comment by whyrusleeping** [comment]
yeah, i'm not really sure. yay race conditionssss

File: exchange/bitswap/bitswap.go [hunk 1]
```diff
@@ -94,6 +86,15 @@ func New(parent context.Context, p peer.ID, network bsnet.BitSwapNetwork,
...
 	// Start up bitswaps async worker routines
 	bs.startWorkers(px, ctx)
+
+	// bind the context and process.
+	// do it over here to avoid closing before all setup is done.
+	go func() {
+		<-px.Closing() // process closes first
+		cancelFunc()
+	}()
+	procctx.CloseAfterContext(px, ctx) // parent cancelled first
```
Comment (rht) [inline exchange/bitswap/bitswap.go]: What's the reason that `bitswap.New` has WithCancel but `dht.NewDHT` doesn't?

File: exchange/bitswap/bitswap.go [hunk 2]
```diff
@@ -94,6 +86,15 @@ func New(parent context.Context, p peer.ID, network bsnet.BitSwapNetwork,
 
 	// Start up bitswaps async worker routines
 	bs.startWorkers(px, ctx)
+
+	// bind the context and process.
+	// do it over here to avoid closing before all setup is done.
+	go func() {
```
Comment (rht) [inline exchange/bitswap/bitswap.go]: Should https://github.com/ipfs/go-ipfs/blob/master/p2p/net/swarm/swarm_dial.go#L420 has this as well?

**Pull request merged**